File: src/main/java/jugglinglab/jml/JMLPosition.java
//
// JMLPosition.java
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.jml;

import java.io.*;
import java.util.*;
import jugglinglab.util.*;

public class JMLPosition {
  static final ResourceBundle guistrings = jugglinglab.JugglingLab.guistrings;
  static final ResourceBundle errorstrings = jugglinglab.JugglingLab.errorstrings;

  protected double x, y, z, t, angle;
  protected int juggler;
  protected JMLPosition prev, next; // for doubly-linked event list

  public JMLPosition() {}

  public Coordinate getCoordinate() {
    return new Coordinate(x, y, z);
  }

  public void setCoordinate(Coordinate c) {
    x = c.x;
    y = c.y;
    z = c.z;
  }

  public double getAngle() {
    return angle;
  }

  public void setAngle(double angle) {
    this.angle = angle;
  }

  public double getT() {
    return t;
  }

  public void setT(double t) {
    this.t = t;
  }

  public int getJuggler() {
    return juggler;
  }

  public void setJuggler(String strjuggler) {
    juggler = Integer.parseInt(strjuggler);
  }

  public void setJuggler(int j) {
    juggler = j;
  }

  public JMLPosition getPrevious() {
    return prev;
  }

  public void setPrevious(JMLPosition prev) {
    this.prev = prev;
  }

  public JMLPosition getNext() {
    return next;
  }

  public void setNext(JMLPosition next) {
    this.next = next;
  }

  public int getHashCode() {
    return toString().hashCode();
  }

  //----------------------------------------------------------------------------
  //  Reader/writer methods
  //----------------------------------------------------------------------------

  public void readJML(JMLNode current, String jmlvers) throws JuggleExceptionUser {
    JMLAttributes at = current.getAttributes();
    double tempx = 0, tempy = 0, tempz = 0, tempt = 0, tempangle = 0;
    String jugglerstr = "1";

    try {
      for (int i = 0; i < at.getNumberOfAttributes(); i++) {
        // System.out.println("att. "+i+" = "+at.getAttributeValue(i));
        if (at.getAttributeName(i).equalsIgnoreCase("x")) {
          tempx = JLFunc.parseDouble(at.getAttributeValue(i));
        } else if (at.getAttributeName(i).equalsIgnoreCase("y")) {
          tempy = JLFunc.parseDouble(at.getAttributeValue(i));
        } else if (at.getAttributeName(i).equalsIgnoreCase("z")) {
          tempz = JLFunc.parseDouble(at.getAttributeValue(i));
        } else if (at.getAttributeName(i).equalsIgnoreCase("t")) {
          tempt = JLFunc.parseDouble(at.getAttributeValue(i));
        } else if (at.getAttributeName(i).equalsIgnoreCase("angle")) {
          tempangle = JLFunc.parseDouble(at.getAttributeValue(i));
        } else if (at.getAttributeName(i).equalsIgnoreCase("juggler")) {
          jugglerstr = at.getAttributeValue(i);
        }
      }
    } catch (NumberFormatException nfe) {
      throw new JuggleExceptionUser(errorstrings.getString("Error_position_coordinate"));
    }

    setCoordinate(new Coordinate(tempx, tempy, tempz));
    setT(tempt);
    setAngle(tempangle);
    if (jugglerstr == null) {
      throw new JuggleExceptionUser(errorstrings.getString("Error_position_nojuggler"));
    }
    setJuggler(jugglerstr);

    if (current.getNumberOfChildren() != 0) {
      throw new JuggleExceptionUser(errorstrings.getString("Error_position_subtag"));
    }
  }

  public void writeJML(PrintWriter wr) throws IOException {
    Coordinate c = getCoordinate();
    wr.println(
        "<position x=\""
            + JLFunc.toStringRounded(c.x, 4)
            + "\" y=\""
            + JLFunc.toStringRounded(c.y, 4)
            + "\" z=\""
            + JLFunc.toStringRounded(c.z, 4)
            + "\" t=\""
            + JLFunc.toStringRounded(getT(), 4)
            + "\" angle=\""
            + JLFunc.toStringRounded(getAngle(), 4)
            + "\" juggler=\""
            + Integer.toString(getJuggler())
            + "\"/>");
  }

  // java.lang.Object methods

  @Override
  public String toString() {
    StringWriter sw = new StringWriter();
    try {
      writeJML(new PrintWriter(sw));
    } catch (IOException ioe) {
    }

    return sw.toString();
  }
}

File: src/main/kotlin/jugglinglab/jml/JMLPosition.kt
//
// JMLPosition.kt
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.jml

import jugglinglab.JugglingLab.errorstrings
import jugglinglab.util.Coordinate
import jugglinglab.util.JuggleExceptionUser
import jugglinglab.util.parseDouble
import jugglinglab.util.toStringRounded
import java.io.IOException
import java.io.PrintWriter
import java.io.StringWriter

class JMLPosition {
    var x: Double = 0.0
    var y: Double = 0.0
    var z: Double = 0.0
    @JvmField var t: Double = 0.0
    @JvmField var angle: Double = 0.0
    var juggler: Int = 0

    var previous: JMLPosition? = null
    var next: JMLPosition? = null // for doubly-linked event list

    var coordinate: Coordinate
        get() = Coordinate(x, y, z)
        set(c) {
            x = c.x
            y = c.y
            z = c.z
        }

    fun setJuggler(strjuggler: String) {
        juggler = strjuggler.toInt()
    }

    val hashCode: Int
        get() = toString().hashCode()

    //--------------------------------------------------------------------------
    //  Reader/writer methods
    //--------------------------------------------------------------------------

    @Suppress("unused")
    @Throws(JuggleExceptionUser::class)
    fun readJML(current: JMLNode, jmlvers: String?) {
        val at = current.attributes
        var tempx = 0.0
        var tempy = 0.0
        var tempz = 0.0
        var tempt = 0.0
        var tempangle = 0.0
        var jugglerstr = "1"

        try {
            for (i in 0..<at.numberOfAttributes) {
                // System.out.println("att. "+i+" = "+at.getAttributeValue(i));
                if (at.getAttributeName(i).equals("x", ignoreCase = true)) {
                    tempx = parseDouble(at.getAttributeValue(i))
                } else if (at.getAttributeName(i).equals("y", ignoreCase = true)) {
                    tempy = parseDouble(at.getAttributeValue(i))
                } else if (at.getAttributeName(i).equals("z", ignoreCase = true)) {
                    tempz = parseDouble(at.getAttributeValue(i))
                } else if (at.getAttributeName(i).equals("t", ignoreCase = true)) {
                    tempt = parseDouble(at.getAttributeValue(i))
                } else if (at.getAttributeName(i).equals("angle", ignoreCase = true)) {
                    tempangle = parseDouble(at.getAttributeValue(i))
                } else if (at.getAttributeName(i).equals("juggler", ignoreCase = true)) {
                    jugglerstr = at.getAttributeValue(i)
                }
            }
        } catch (_: NumberFormatException) {
            throw JuggleExceptionUser(errorstrings.getString("Error_position_coordinate"))
        }

        coordinate = Coordinate(tempx, tempy, tempz)
        t = tempt
        angle = tempangle
        setJuggler(jugglerstr)

        if (current.numberOfChildren != 0) {
            throw JuggleExceptionUser(errorstrings.getString("Error_position_subtag"))
        }
    }

    @Throws(IOException::class)
    fun writeJML(wr: PrintWriter) {
        val c = this.coordinate
        wr.println(
            ("<position x=\""
                    + toStringRounded(c.x, 4)
                    + "\" y=\""
                    + toStringRounded(c.y, 4)
                    + "\" z=\""
                    + toStringRounded(c.z, 4)
                    + "\" t=\""
                    + toStringRounded(this.t, 4)
                    + "\" angle=\""
                    + toStringRounded(this.angle, 4)
                    + "\" juggler=\""
                    + this.juggler
                    + "\"/>")
        )
    }

    override fun toString(): String {
        val sw = StringWriter()
        try {
            writeJML(PrintWriter(sw))
        } catch (_: IOException) {
        }
        return sw.toString()
    }
}
